client.go

```diff
@@ -7,83 +7,155 @@ package bitmarkdClient
 import (
 	"crypto/tls"
 	"fmt"
+	"net"
 	"net/rpc"
 	"net/rpc/jsonrpc"
 	"sync"
+	"sync/atomic"
 	"time"
 )
 
-// ErrRPCConnection is an error for rpc connection
-var ErrRPCConnection = fmt.Errorf("can not connect to rpc server")
+var ErrRPCRequestTimeout = fmt.Errorf("rpc requests timed out")
+var ErrRPCConnectionClosed = fmt.Errorf("rpc connection closed")
+var ErrCannotEstablishConnection = fmt.Errorf("connection can not be established")
 
 // RPCEmptyArguments is an empty argument for rpc requests
 type RPCEmptyArguments struct{}
 
-// BitmarkdRPCClient is a struct for bitmarkd
+// PersistentRPCClient is client that will maintain a long-lived connection for requests
+type PersistentRPCClient struct {
+	sync.RWMutex
+	*rpc.Client
+	address   string
+	closed    chan struct{}
+	connected bool
+	timeout   time.Duration
+}
+
+func NewPersistentRPCClient(address string, timeout time.Duration) *PersistentRPCClient {
+	return &PersistentRPCClient{
+		address: address,
+		timeout: timeout,
+		closed:  make(chan struct{}),
+	}
+}
+
+// Call is to make an RPC request. It will whether the connection is still alived.
+// If not, it will try to create one.
+func (c *PersistentRPCClient) Call(serviceMethod string, args interface{}, reply interface{}) error {
+	if !c.connected {
+		if err := c.connect(); err != nil {
+			return ErrCannotEstablishConnection
+		}
+	}
+
+	c.RLock()
+	defer c.RUnlock()
+	if !c.connected {
```

Comment:
`c.connected` could be false if `connect()` doesn't return error?

Reply:
There are race conditions between locks. I will  remove the lock and ask the connection function to return a client immediately. The rest of case will be left to the select block.